### ISSUE TITLE
Incorrect cross-platform file path resolution with include_str!

### ISSUE THREAD

**Issue by kevinpoitra** [rust-lang/rust#55833]
Last night, I updated my local copy of Rust (stable-x86_64-windows-gnu), and updated the dependencies within my project (which mainly consists of GTK/Relm related libraries). The file structure of my project is like so:

- src
  - assets
    - style.css
  - views
    - view1.rs
    - view2.rs
    - ...
  - main.rs

The project itself is built in two ways: Locally, on my Windows 10 machine, and via CI to cross-compile it for a Raspberry Pi (running Raspbian, using target armv7-unknown-linux-gnueabihf). After updating, it seems like `include_str!`'s file path resolution logic doesn't match across the targets, seeing as my project was successfully building on both targets yesterday. To be more clear, after updating, this code:

`include_str!("assets/style.css")`

which is called from within any of the viewx.rs files, will output the following on Windows:

`couldn't read src\views\assets/style.css: The system cannot find the path specified. (os error 3)`

but will compile just fine for the Pi. Changing the path to `../assets/style.css` will allow it to compile on Windows, but will break the Pi's build, with an error of:

`error: couldn't read src/../assets/style.css: No such file or directory (os error 2)`

I'm not sure what `include_str!` is looking for in terms of path now, as previously 'assets/style.css' correctly pointed to the file on both targets (and by previously I mean likely Rust 1.29 - I also switched from nightly to stable but the error persists on both). [The docs for `include_str!`](https://doc.rust-lang.org/std/macro.include_str.html) say that the path should be relative to the 'current file'. It seems that on Windows, the 'current file' is the file that's calling `include_str!` (i.e. view1.rs), whereas on the Pi/Linux, it's main.rs itself.

**Comment by retep998** [comment]
Somewhere a path has been canonicalized into a `\\?\` path, and appending a relative path that uses `/` onto that will break because `\\?\` paths do *not* interpret `/` as a path separator.

**Comment by kevinpoitra** [comment]
Yeah that seems to be it, or related to it in some way. For what it's worth, I did some more digging:

On Windows:
- `assets/style.css` outputs `couldn't read src\views\assets/style.css: The system cannot find the path specified. (os error 3)`
- `assets\\style.css` outputs `couldn't read src\views\assets\style.css: The system cannot find the path specified. (os error 3)`
- `\\assets\\style.css` outputs `couldn't read \assets\style.css: The system cannot find the path specified. (os error 3)`

On Raspbian:
- `assets/style.css` lets the program successfully build.
- `assets\\style.css` outputs `error: couldn't read src/assets\style.css: No such file or directory (os error 2)`
- `\\assets\\style.css` outputs `error: couldn't read src/\assets\style.css: No such file or directory (os error 2)`

On top of that, when the Windows build errors out, the code that it points to is the call to `include_str!` itself. On Raspbian, when it errors out, it points to where my program calls Relm's `Win::run` function within main.rs.

Additionally, moving the `assets` folder into the `views` folder makes it so the first two paths (`assets/style.css` and `assets\\style.css`) can compile on Windows, but breaks the Pi's build for any path format.

**Comment by kevinpoitra** [comment]
I don't think this is an error with Rust itself anymore, so I'll close this for now.

Based on my last two observations from my previous comment, I originally thought there was something wrong around [this line within the res_rel_file function](https://github.com/rust-lang/rust/blob/25a42b2ceb46887e9941cec667eac99844dd7ad0/src/libsyntax/ext/source_util.rs#L201), used by `include_str!`. With no recent changes to it though, I decided to look elsewhere.

As mentioned previously, my program uses the Relm crate. I was using Relm's `#[widget]` macro to create my UIs, which relies on `relm-attributes`. I downgraded the library versions within my program and downgraded Rust itself, but the error persisted. At this point, I'm not sure how the program was building correctly in the first place. What ended up working was changing my program's view files to not be based on Relm's `#[widget]` macro. In doing so, `assets/style.css` with the original file structure can be corrected resolved on both Windows and Raspbian.
